#[130388891] Admin should be able to review and on-board artisans

hmasila opened the pull request:
#### What does this PR do?

create admin `user_type`
create status column in the `user` model
define menu options for admin `user_type`
create admin view to list applicants
create admin view to assess applicant responses and either rejecting or accepting the applicant
#### What are the relevant pivotal tracker stories?

130388891
#### Screenshots

![image](https://cloud.githubusercontent.com/assets/20906768/18934636/968ac964-85e3-11e6-8952-b85b1508ba2f.png)
![image](https://cloud.githubusercontent.com/assets/20906768/18934646/a7cfbea0-85e3-11e6-8a03-e5800f8df726.png)

Commit:
feat(application-review): review artisan application

create admin `user_type`
create status column in the user model
define menu options for admin user_type
create admin view to list applicants
create admin view to assess applicant responses and either rejecting or accepting the applicant

Commit:
bg(routes): change the routes from admin/applications .. to admin/view_applications ..

andela-fadebayo commented on app/assets/stylesheets/styles/admin.scss:
```diff
@@ -0,0 +1,33 @@
+
+.container-large {
+  padding-left: 20em;
```
For consistency, you should stick to one variable naming convention. It's preferable to use hyphens "-". Ensure you use this across all classes and ids.

hmasila replied:
Do you mean hyphens(-) or underscores(_)? I have looked at the other scss in the project and they are using underscores

andela-fadebayo commented on app/assets/stylesheets/styles/admin.scss:
```diff
@@ -0,0 +1,33 @@
+.trade{
+  display: inline;
+}
+.noScroll {
```
Use snake-case here.

hmasila replied:
noted

andela-fadebayo commented on app/assets/stylesheets/styles/admin.scss:
```diff
@@ -0,0 +1,33 @@
+  margin-bottom: auto;
+}
+
+.btn_flat{
```
Ensure to follow consistency in spacing between classes and curly braces.

andela-hmasila replied:
👍

andela-fadebayo commented on app/assets/stylesheets/styles/admin.scss:
```diff
@@ -0,0 +1,33 @@
+}
+.answers {
+  border-bottom: 1px solid;
+  margin: 20px
```
Missed semicolon.

andela-fadebayo commented on app/controllers/admin_controller.rb:
```diff
@@ -0,0 +1,25 @@
+class AdminController < ApplicationController
```
It's a best practice to have the class naming agree with what actions are performed. Since these are actions of an admin on applicants, this class could be `Admin::ApplicantsController`.

hmasila replied:
This brings issues with the templates. How should the views be named in this case?

andela-fadebayo commented on app/controllers/admin_controller.rb:
```diff
@@ -0,0 +1,25 @@
+  before_action :login_required
+  before_action :check_admin
+
+  def view_applications
```
These controller actions should follow Rails RESTful protocol. Since you're returning all admin applicants, this should be the `index` action.

andela-fadebayo commented on app/controllers/admin_controller.rb:
```diff
@@ -0,0 +1,25 @@
+    @applicants = Admin.get_applicants
+  end
+
+  def review_applicant
```
This should be `show` action.

hmasila replied:
I think it's `edit` action because at the end there is a form that edits the status of the applicant. The applicant is either rejected or accepted

andela-fadebayo commented on app/controllers/admin_controller.rb:
```diff
@@ -0,0 +1,25 @@
+    @responses = Response.where("user_id = ?", params[:id])
+  end
+
+  def update_applicant
```
And `update` action.

andela-fadebayo commented on app/controllers/admin_controller.rb:
```diff
@@ -0,0 +1,25 @@
+    redirect_to(applications_path)
+  end
+
+  def check_admin
```
Seems okay.

andela-fadebayo commented on app/models/admin.rb:
```diff
@@ -0,0 +1,5 @@
+class Admin < User
+  def self.get_applicants
+    User.where(["user_type = ? and status = ?", "taskee", 0])
```
This is a query fetching applicants. Use `scope` instead.

andela-fadebayo commented on app/views/admin/review_applicant.html.erb:
```diff
@@ -0,0 +1,51 @@
+    <div class="row">
+        <div class="modal-content col s12">
+            <h4><%= @applicant.fullname%></h4>
+            <ul class="li_item">
```
Use snake-case with hyphen `-` in class names for consistency.

andela-fadebayo commented on app/views/admin/review_applicant.html.erb:
```diff
@@ -0,0 +1,51 @@
+                <% end %>
+            </ul>
+
+            <form action="<%= update_applicant_path %>" method="post">
```
Following my comment on the controller, you can use Rails form helpers here with the `update` action.

andela-fadebayo commented on app/views/admin/view_applications.html.erb:
```diff
@@ -0,0 +1,26 @@
+            </ul>
+            <p>
+                Submitted on:
+                <%= Date.strptime(applicant.updated_at.to_s, "%Y-%m-%d")%>
```
Move this logic from the view to a helper method.

andela-fadebayo commented on config/routes.rb:
```diff
@@ -2,6 +2,10 @@
+  get "admin/dashboard" => "dashboard#home", as: "admin_dashboard"
+  get "admin/view_applications" => "admin#view_applications", as: "applications"
+  get "admin/view_applications/:id/review_applicant" => "admin#review_applicant", as: "review_applicant"
+  post "admin/view_applications/:id" => "admin#update_applicant", as: "update_applicant"
```
Why are we not using Rails standard RESTful routing (also using resources) here?

andela-fadebayo commented on spec/controllers/admin_controller_spec.rb:
```diff
@@ -0,0 +1,10 @@
+    it "returns http success" do
+      get :applications
+      expect(response).to have_http_status(:success)
+    end
```
This is good, but we need to be also testing the actual response.

andela-fadebayo commented on spec/factories/responses.rb:
```diff
@@ -0,0 +1,5 @@
+FactoryGirl.define do
+  factory :response do
+    response ""
```
How does `response` look like?

andela-fadebayo commented on spec/helpers/admin_helper_spec.rb:
```diff
@@ -0,0 +1,15 @@
+# end
+RSpec.describe AdminHelper, type: :helper do
+  pending "add some examples to (or delete) #{__FILE__}"
+end
```
This file should not exist if not used. Also remove comments.

andela-fadebayo commented on spec/models/response_spec.rb:
```diff
@@ -0,0 +1,5 @@
+require "rails_helper"
+
+RSpec.describe Response, type: :model do
+  pending "add some examples to (or delete) #{__FILE__}"
```
Add tests.

Commit:
ch(refactor): refactor code with recommendations from PR
- write tests
-use nested resources
- update routes
- add reason to user table
- use RESTful actions

Commit:
bg(views): make radio button in review applicant to toggle
- use strong parameters for the applicants controller
- update the radio button ids to match the label ids
- write tests for the updates
- update user_attr in tests to include status and reason
- update completion percentages caused by the new attributes

andela-fadebayo commented on spec/support/helpers.rb:
```diff
@@ -18,11 +18,13 @@ def taskee_stub
-    @statuses = %w(done active inactive rejected)
-
     @tasker = create(:user, user_type: "tasker")
     @taskee = create(:user, taskee_attr)
```
Change all instance of `taskee` here to `artisan`

andela-fadebayo commented on spec/controllers/dashboard_controller_spec.rb:
```diff
@@ -4,15 +4,20 @@
+                      has_taken_quiz: true,
+                      status: "accepted",
+                      reason: "good")
+      end
```
Upadate FactoryGirl with these new fields.

Commit:
ch(naming): rename all tasekees to artisans
- rename all the taskees and taskee_occurences to artisans
- write passing tests for the review applicant radio buttons

Commit:
bg(travis): add  to before script

Commit:
bg(travis): add  to before script

Commit:
bg(travis): add postgres to services

Commit:
bg(travis): remove migrations in before_install

Commit:
bg(travis): update the selenium link to a working link

Commit:
bg(travis): update the selenium link to a working link
